abstract_domains/numerical/linear_forms.py

```diff
@@ -0,0 +1,186 @@
+from core.expressions import *
+
+from core.expressions_tools import ExpressionVisitor
+from core.special_expressions import VariadicArithmeticOperation
+
+Sign = UnaryArithmeticOperation.Operator
+PLUS = Sign.Add
+MINUS = Sign.Sub
+
+
+class InvalidFormError(ValueError):
+    pass
+
+
+class LinearForm:
+    """Holds an expression in linear form with one or several variables: `+/- var1 +/- var2 + ... + interval`."""
+
+    def __init__(self, expr: Expression):
+        """Initializes this instance with the linear form of an expression.
+
+        If possible, this instance holds the parts of the linear form separately. If not possible to 
+        construct this form, this raises a InvalidFormError.
+        """
+        self._var_summands = {}  # dictionary holding {var: sign}
+        self._interval = None
+
+        LinearForm._visitor.visit(expr, self)
+
+    @property
+    def var_summands(self):
+        return self._var_summands
+
+    def _encounter_new_var(self, var, sign=PLUS):
+        if var in self.var_summands:
+            raise InvalidFormError(f"VariableIdentifier {var} appears twice!")
+        self.var_summands[var] = sign
+
+    @property
+    def interval(self):
+        return self._interval
+
+    @interval.setter
+    def interval(self, value):
+        if self._interval:
+            raise InvalidFormError("interval set twice (is immutable)!")
+        self._interval = value
+
+    def __eq__(self, other: 'LinearForm'):
+        return isinstance(other, self.__class__) \
+               and set(self.var_summands.keys()) == set(other.var_summands.keys()) \
+               and self.interval == other.interval
+
+    def __ne__(self, other: 'LinearForm'):
+        return not (self == other)
+
+    def __lt__(self, other):
+        """Syntactic comparision of this linear form."""
+        if not isinstance(other, self.__class__):
+            raise NotImplementedError("Incomparable types!")
+        return self.var_summands == other.var_summands \
+               and self.interval < other.interval
+
+    def __le__(self, other):
+        """Syntactic comparision of this linear form."""
+        if not isinstance(other, self.__class__):
+            raise NotImplementedError("Incomparable types!")
+        return self.var_summands == other.var_summands \
+               and self.interval <= other.interval
+
+    def __gt__(self, other):
+        """Syntactic comparision of this linear form."""
+        if not isinstance(other, self.__class__):
+            raise NotImplementedError("Incomparable types!")
+        return self.var_summands == other.var_summands \
+               and self.interval > other.interval
+
+    def __ge__(self, other):
+        """Syntactic comparision of this linear form."""
+        if not isinstance(other, self.__class__):
+            raise NotImplementedError("Incomparable types!")
+        return self.var_summands == other.var_summands \
+               and self.interval >= other.interval
```

Comment:
I guess these are the ones using the comparisons in Interval. But the we need these? And again, are we sure about the semantics?